src/validator/validator.ts

```diff
@@ -58,7 +56,8 @@ export abstract class VObjectBase<T extends Schema> {
           keys.push(...v.keys)
           walk(v.container as T, keys, isOptional)
           const tmp: string[] = []
-          Object.assign(tmp, thisKeys)
+          const lastObjectInArray = v instanceof VArray && keys[keys.length - 2] === '[*]'
+          Object.assign(tmp, keys.slice(0, keys.length - (lastObjectInArray ? 3 : 1)))
```

Comment:
@yusukebe the issue here was that we were forgetting the keys after finishing the check of an object, for example in:
```ts
    const vObject = v.object('blog', v => ({
      posts: v.array('posts', v => ({
        author: v.object('author', v => ({
          name: v.json('name'),
        })),
        publisher: v.object('publisher', v => ({
          name: v.json('name'),
        }))
      })),
    }))
```
after having walked through the author, the keys would be reset to `['blog']` (instead of `['blog', 'posts']`) so I figured that we shouldn't always use the same `thisKeys` but always use the keys and just push and pop values as needed.

the removal is for 1 element if the last element was not the last in an array (for example in `blog.posts.[*].publisher.name` we want to remove `name`) otherwise it is 3 because we also need to remove the `[*]` and the name of the array (for example in `blog.posts.[*].publisher` we want to remove `publisher`, `[*]` and `posts`) 

What do you think? does this solution make sense to you? :thinking:

Reply:
Hi. If we fix the positions of the array, it seems the test below failed. Or not?

```ts
describe.only('Nested objects and arrays', () => {
  const blogJson = {
    blog: {
      meta: {
        foo: {
          bar: 'bar',
        },
      },
      posts: [
        {
          author: {
            name: 'foo',
          },
        },
      ],
    },
  }

  it('Should correctly handle values inside objects and arrays', async () => {
    const req = new Request('http://localhost/', {
      method: 'POST',
      body: JSON.stringify(blogJson),
    })

    const v = new Validator()
    const vObject = v.object('blog', (v) => ({
      meta: v.object('meta', (v) => ({
        foo: v.object('foo', (v) => ({
          bar: v.json('bar'),
        })),
      })),
      posts: v.array('posts', (v) => ({
        author: v.object('author', (v) => ({
          name: v.json('name'),
        })),
      })),
    }))
    for (const validator of vObject.getValidators()) {
      const results = await validator.validate(req)
      expect(results[0].isValid).toBe(true)
    }
  })
})
```

Reply:
@yusukebe my solution was a bit silly, I've updated it to what I believe to be a correct solution, I'll mark the PR as ready to review, please have a look :slightly_smiling_face:

Reply:
Wow, cool implementation! Looks good!